### PR TITLE
Introduce event aggregation on AWS

### PR DESCRIPTION
**Pull request by gbossert**
This PR introduces the notion of event aggregation.

Events that share a comment fingerprint will be aggregated under a TTL strategy.

### CODE REVIEW

File: AWS/aws/event_aggregator.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+class Aggregation(BaseModel):
+    start: datetime.datetime
+    end: datetime.datetime
+    count: int
+    event: dict
+
+    def get_aggregated_event(self):
+        aggregated_event = self.event
+        aggregated_event["event"]["start"] = self.start.isoformat()
+        aggregated_event["event"]["end"] = self.start.isoformat()
```
Comment (Darkheir) [inline AWS/aws/event_aggregator.py:24]:
```suggestion
        aggregated_event["event"]["end"] = self.end.isoformat()
```

File: AWS/aws/event_aggregator.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+            event_hash = self.get_hash(event)
+            # if no hash can be computed, we don't aggregate and forward the event
+            if not event_hash:
+                return event
+
+            # the aggregation stores the parsed timestamp
+            event_dt = parse_datetime_as_naive(event["@timestamp"])
+
+            # if hash is already known
+            with self.lock:
```
Comment (Darkheir) [inline AWS/aws/event_aggregator.py:145]: So if I understand correctly, first occurrence is sent immediately and the aggregated one will follow once TTL is reached. Is that right ?

File: AWS/aws/event_aggregator.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+class EventAggregator:
+    ttl_thread: EventAggregatorTTLThread | None
+    lock: threading.Lock
+
+    def __init__(self, aggregation_definitions: dict[str, list[Fingerprint]]):
+        self.aggregation_definitions = aggregation_definitions
+        self.aggregations: dict[str, Aggregation] = dict()
+        self.lock = threading.Lock()
+
+    def start_flush_on_ttl(self, ttl: int, on_flush_func: Callable[[dict], None], delay: float = 10):
```
Comment (Darkheir) [inline AWS/aws/event_aggregator.py:71]: TTL is global and can't be changed depending on the kind of event. 

I'm totally fine with it, but I'm wondering if we can have cases where some aggregations should be forwarded faster than other ?

File: AWS/aws/event_aggregator.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        with self.lock:
+            for event_hash, aggregation in self.aggregations.items():
+                if aggregation.start + datetime.timedelta(seconds=ttl) < datetime.datetime.utcnow():
+                    event_hashes_to_flush.append(event_hash)
+
+                    if aggregation.count > 0:
+                        aggregated_events.append(aggregation.get_aggregated_event())
+
+            for event_hash in event_hashes_to_flush:
+                del self.aggregations[event_hash]
```
Comment (Darkheir) [inline AWS/aws/event_aggregator.py:119]: Could be done in the first iteration since we iterate on `.items()`

**Review by Darkheir — COMMENTED** [review]
Code logic looks good.

It would be nice to have it in the SDK so all the connectors may benefit of this capability

File: AWS/aws/event_aggregator.py (hunk 2 above)
Comment (gbossert) [reply AWS/aws/event_aggregator.py:145]: yes exactly

File: AWS/aws/event_aggregator.py (hunk 3 above)
Comment (gbossert) [reply AWS/aws/event_aggregator.py:71]: well spotted, I will upgrade the aggregation definition to support custom ttl

File: AWS/aws/event_aggregator.py (hunk 4 above)
Comment (gbossert) [reply AWS/aws/event_aggregator.py:119]: ok I didn't know this method was compatible with dict edition. I will simplify accordingly.

**Comment by gbossert** [comment]
This PR is moved in the SDK : https://github.com/SEKOIA-IO/sekoia-automation-sdk/pull/87

File: AWS/aws/event_aggregator.py (hunk 4 above)
Comment (gbossert) [reply AWS/aws/event_aggregator.py:121]: .items() doesnt seems compatible with inplace edition of dict. I used list(.keys()) instead

**Pull request closed without merging**